pkg/agent/bgp/gobgp/gobgp.go

```diff
@@ -0,0 +1,288 @@
+	} else if routeType == bgp.RouteReceived {
+		gobgpTableType = gobgpapi.TableType_ADJ_IN
+	} else {
+		return nil, fmt.Errorf("invalid route type: %v", routeType)
+	}
+	if peerAddress == "" {
+		return nil, fmt.Errorf("BGP peer IP address must be specified")
+	}
+
+	var routes []bgp.Route
+	fn := func(destination *gobgpapi.Destination) {
+		if destination == nil {
+			return
+		}
+		routes = append(routes, bgp.Route{Prefix: destination.GetPrefix()})
+	}
+
+	request := &gobgpapi.ListPathRequest{
+		TableType: gobgpTableType,
+		Family:    &gobgpapi.Family{Afi: toGoBGPFamilyAfi(net.IsIPv6String(peerAddress)), Safi: gobgpapi.Family_SAFI_UNICAST},
+		Name:      peerAddress,
+	}
+
+	if err := s.server.ListPath(context.TODO(), request, fn); err != nil {
+		return nil, err
+	}
+	return routes, nil
+}
+
+func toGoBGPPath(route *bgp.Route) *gobgpapi.Path {
+	isIPv6 := net.IsIPv6CIDRString(route.Prefix)
+	goBGPIPFamily := toGoBGPFamilyAfi(isIPv6)
+	prefix, _ := netip.ParsePrefix(route.Prefix)
+	nlri, _ := anypb.New(&gobgpapi.IPAddressPrefix{
+		Prefix:    prefix.Addr().String(),
+		PrefixLen: uint32(prefix.Bits()),
+	})
+
+	var attrs []*anypb.Any
+	a1, _ := anypb.New(&gobgpapi.OriginAttribute{
+		Origin: 0,
+	})
+	var a2 *anypb.Any
+	if isIPv6 {
+		a2, _ = anypb.New(&gobgpapi.MpReachNLRIAttribute{
+			Family:   &gobgpapi.Family{Afi: goBGPIPFamily, Safi: gobgpapi.Family_SAFI_UNICAST},
+			NextHops: []string{ipv6AllZero},
+			Nlris:    []*anypb.Any{nlri},
+		})
+	} else {
+		a2, _ = anypb.New(&gobgpapi.NextHopAttribute{
+			NextHop: ipv4AllZero,
+		})
+	}
+	attrs = append(attrs, a1, a2)
+
+	return &gobgpapi.Path{
+		Family: &gobgpapi.Family{Afi: goBGPIPFamily, Safi: gobgpapi.Family_SAFI_UNICAST},
+		Nlri:   nlri,
+		Pattrs: attrs,
+	}
+}
+
+func toGoBGPFamilyAfi(isIPv6 bool) gobgpapi.Family_Afi {
+	if isIPv6 {
+		return gobgpapi.Family_AFI_IP6
+	}
+	return gobgpapi.Family_AFI_IP
+}
+
+func toGoBGPPeer(peerConfig bgp.PeerConfig) *gobgpapi.Peer {
+	peer := &gobgpapi.Peer{
+		Conf: &gobgpapi.PeerConf{
+			NeighborAddress: peerConfig.Address,
+			PeerAsn:         uint32(peerConfig.ASN),
+			AuthPassword:    peerConfig.Password,
+		},
+		AfiSafis: []*gobgpapi.AfiSafi{
+			{
+				Config: &gobgpapi.AfiSafiConfig{
+					Family: &gobgpapi.Family{
+						Afi:  toGoBGPFamilyAfi(net.IsIPv6String(peerConfig.Address)),
+						Safi: gobgpapi.Family_SAFI_UNICAST,
+					},
+					Enabled: true,
+				},
+				MpGracefulRestart: &gobgpapi.MpGracefulRestart{
+					Config: &gobgpapi.MpGracefulRestartConfig{
+						Enabled: true,
+					},
+				},
+			},
+		},
+	}
+	if peerConfig.Port != nil {
+		peer.Transport = &gobgpapi.Transport{
+			RemotePort: uint32(*peerConfig.Port),
+		}
+	}
+	if peerConfig.MultihopTTL != nil && *peerConfig.MultihopTTL != bgp.DefaultBGPMultihopTTL {
+		peer.EbgpMultihop = &gobgpapi.EbgpMultihop{
+			Enabled:     true,
+			MultihopTtl: uint32(*peerConfig.MultihopTTL),
+		}
+	}
+	if peerConfig.GracefulRestartTimeSeconds != nil {
```

Comment:
Unless I am mistaken, `peerConfig.Port`, `peerConfig.MultihopTTL`, `peerConfig.GracefulRestartTimeSeconds` will always be set at this point, because the default value will be populated by the apiserver? That being said it may be good to keep the `nil` checks. cc @tnqn

Reply:
> Unless I am mistaken, peerConfig.Port, peerConfig.MultihopTTL, peerConfig.GracefulRestartTimeSeconds will always be set at this point, because the default value will be populated by the apiserver? 

You are right, and it seems that it is a convention to do the nil checks even when the pointer fields have default values (not very sure about that).

Reply:
Personally I would prefer writing code more precisely, i.e. if it should never be nil, then do not handle nil case. It helps making code clean and improves test coverage. For user inputs or values got from external system which is not under control, I would prefer validating them in the first place (e.g. the public functions) and skip validating them repeatedly in private functions.

But I see such error or nil check often too. If it's harmless, I don't have strong opinion.

Reply:
BTW, if these values are really nil, and we skip setting the corresponding gobpp parameters, will there be a problem?

Reply:
I believe that if these values are `nil`, gobgp will use the default behavior (default transport port value, multihop disabled, graceful restart disabled).
I think the current version of the code is ok. But we could also return an error in `AddPeer` / `UpdatePeer` if any required field is missing. The "problem" is that the interface here is tightly coupled to the `BGPPolicy` CRD. When a CR is created through the apiserver, all these fields should be defaulted and should not be `nil`, but is it a good thing to rely on this when implementing this interface, even if the `BGPPolicy` controller will be the only consumer of the interface?
I am fine with returning an error if any of these fields is `nil`. Realistically , this only matters if we have multiple implementations of the interface, and we want to try to achieve the same behavior across all of them, without having to introduce duplicate defaulting logic here (as it should always be provided by the CRD).

Reply:
> BTW, if these values are really nil, and we skip setting the corresponding gobpp parameters, will there be a problem?

As @antoninbas said, gobgp will use the default values. The corresponding code is here: https://github.com/osrg/gobgp/blob/0148e2d22dcf40655d2b75c6f40bf49a378c4740/pkg/config/oc/default.go#L62.

Reply:
> I think the current version of the code is ok. But we could also return an error in AddPeer / UpdatePeer if any required field is missing. 

Will add the checks for required fields like peer address and ASN.